### PR TITLE
Iqss/8109 separate managefilepermissions

### LINKED ISSUE #8109: Split off GrantFileAccess permission from ManageDatasetPermissions
Problem
======
Currently, the permission to grant file access is grouped together with the permission to assign roles to users and groups under `ManageDatasetPermissions`. This is problematic in the following scenario: 

* You have configured a dataverse so that dataset creators get the role "contributor" on the datasets they create, meaning   that the dataset has to be reviewed and published by a curator.
* You still want to delegate the responsibility to grant access to restricted files to the dataset creator.

To enable the creator to handle the access requests you could assign them a role that includes  `ManageDatasetPermissions`. However, that would also enable them to assign to themselves any other role, including curator, thus enabling them to publish new versions of the dataset without prior review.

Solution
======

By separating `ManageDatasetPermissions` into `ManageDatasetPermissions` (assigning roles) and `GrantFileAccess` (granting access to files) this can be resolved.

Code changes
==========

* A new permission `GrantFileAccess` will be created.
* All permission checks in the code that currently check for `ManageDatasetPermissions` will be reviewed. In the occurrences where the protected action is granting file access, the check's argument will be changed to `GrantFileAccess`.
* A flyway script will be created that updates the table `dataverserole` so that all roles that include the permission bit for `ManageDatasetPermissions` will additionally get the bit for `GrantFileAccess`.

### PR DESCRIPTION
**Pull request by qqmyers**
**What this PR does / why we need it**: 

Adds a new ManageFilePermissions permission defined on datasets that is required to add/revoke FileDownload permissions and view file download requests, etc. 

This then allows defining a role that would allow a user to respond to file download requests without also giving them the power to manage the dataset (which allows you to then increase your own permissions on that dataset.)

Any existing roles with the ManageDatasetPermissions permission are updated to also have this new permission (the existing admin and curator roles  are the only two standard roles that have the ManageDatasetPermissions permission) via a flyway script and the role definitions for admin and curator ( json files in the scripts/api/data folder) are also updated.

**Which issue(s) this PR closes**:

Closes #8109

**Special notes for your reviewer**: Should this PR define a new role that can manage files and not the dataset permissions (i.e. the use case driving the PR) versus having sites create their own custom roles? Any name suggestions?

**Suggestions on how to test this**: 
1)Verify that admins and curators can still add/remove file download permissions (and can do all of the other dataset-level permission assignments as well ).
2) Create a role having the new ManageFilePermissions permission and not the ManageDatasetPermissions, assign a user to that role for a given dataset, and verify that a) that user can see the Permissions/File menu item and not the Permissions/Dataset option, and b) that they can then use the file permissions page to add/revoke file download permissions. Also verify that the user can run the /api/access/datafile/{id}/listRequests and /api/access//datafile/{id}/rejectAccess/{identifier} api calls on that dataset/its files.
3) FWIW: In the db, doing a `select id, perimissionbits::bit(14) from dataverserole;` is a useful way to see the bitmap of permissions assigned to a role. What should be happening in the update it that any role with bit 9 should have a new bit 10 with the same value and the existing bit 10-13 values shifted into bits 11-14. Using letters for bit positions (a=1 through m=13):
mlkjihgfedcba --> mlkjiihgfedcba

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**: 
Changes the Dataset/Edit Dataset/Permissions menu to only have one option - Dataset or File - if you only have the original ManageDatasetPermissions or the new ManageFilePermissions permission, respectively. (Existing roles which have both permissions after the PR still see a menu with the original two options.). 
The new ManageFilePermissions permission shows up in the list for the admin and curator roles where that is shown (e.g. on the permission management dialogs).

**Is there a release notes update needed for this change?**: probably worth mentioning - will wait to draft until we decide whether there should be a new standard role having this permission or not.

**Additional documentation**: tbd - same issue as release notes - will wait for decision re: whether there's a new standard role involved.

### CODE REVIEW

**Comment by djbrooke** [comment]
@scolapasta I know it's still a draft PR but I'll pre-assign you as I know you are the permissions guru.

File: src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AssignRoleCommand.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public RoleAssignment execute(CommandContext ctxt) throws CommandException {
         // for data file check permission on owning dataset
         return Collections.singletonMap("",
                 defPoint instanceof Dataverse ? Collections.singleton(Permission.ManageDataversePermissions)
-                : Collections.singleton(Permission.ManageDatasetPermissions));
+                : defPoint instanceof Dataset ? Collections.singleton(Permission.ManageDatasetPermissions) : Collections.singleton(Permission.ManageFilePermissions));
```
[the remark this answers is not in the document]
Comment (scolapasta) [reply src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AssignRoleCommand.java:79]: @qqmyers let's discuss briefly today and we can decide on the change or not and move along.

File: src/main/java/edu/harvard/iq/dataverse/authorization/Permission.java [hunk 2]
```diff
@@ -46,7 +46,7 @@
     EditDataset(BundleUtil.getStringFromBundle("permission.editDataset"), true, Dataset.class),
     ManageDataversePermissions(BundleUtil.getStringFromBundle("permission.managePermissionsDataverse"), true, Dataverse.class),
     ManageDatasetPermissions(BundleUtil.getStringFromBundle("permission.managePermissionsDataset"), true, Dataset.class),
-    ManageFilePermissions(BundleUtil.getStringFromBundle("permission.managePermissionsDatasetFiles"), true, Dataset.class),
+    ManageFilePermissions(BundleUtil.getStringFromBundle("permission.managePermissionsDataFiles"), true, DataFile.class),
```
Comment (scolapasta) [inline src/main/java/edu/harvard/iq/dataverse/authorization/Permission.java:49]: Very Minor change, but let's make this singular to be in parallel with the others.

File: src/main/java/propertyFiles/Bundle.properties [hunk 3]
```diff
@@ -2216,7 +2216,7 @@ permission.downloadFile=Download a file
 permission.viewUnpublishedDataset=View an unpublished dataset and its files
 permission.viewUnpublishedDataverse=View an unpublished dataverse
 permission.addDatasetDataverse=Add a dataset to a dataverse
-permission.managePermissionsDatasetFiles=Manage permissions for a dataset's files
+permission.managePermissionsDataFiles=Manage permissions for a dataFile(s)
```
Comment (scolapasta) [inline src/main/java/propertyFiles/Bundle.properties]: remove the s and there's an extraneous "a"

File: src/main/java/propertyFiles/Bundle.properties [hunk 4]
```diff
@@ -2535,7 +2535,7 @@ permission.DownloadFile.desc=Download a file
 permission.ViewUnpublishedDataset.desc=View an unpublished dataset and its files
 permission.ViewUnpublishedDataverse.desc=View an unpublished dataverse
 permission.AddDataset.desc=Add a dataset to a dataverse
-permission.ManageFilePermissions.desc=Manage permissions for a dataset's files
+permission.ManageFilePermissions.desc=Manage permissions for a file(s)
```
[the remark this answers is not in the document]
Comment (qqmyers) [reply src/main/java/propertyFiles/Bundle.properties]: Changed the text and moved these next to the dataset/dataverse ones. Note that this line doesn't have an extraneous s (if I understand) - the names of the permissions are plural, but the labels defined for them are a different form and reference a singular object.

**Review by scolapasta — APPROVED** [review]
Changes look good. Thanks.

**Comment by pdurbin** [comment]
> **Is there a release notes update needed for this change?**: probably worth mentioning - will wait to draft until we decide whether there should be a new standard role having this permission or not.
> 
> **Additional documentation**: tbd - same issue as release notes - will wait for decision re: whether there's a new standard role involved.

I saw this went into QA which is great but there are no docs and no release note. From standup yesterday it sounds like we decided that we're not adding a new role in this pull request, which sounds fine. That probably means we don't need any docs. Does it also mean we don't need a release note? Perhaps a release note could at least mention the new ManageFilePermissions permission and indicate UI changes (from above):

> Does this PR introduce a user interface change? If mockups are available, please link/include them here:
Changes the Dataset/Edit Dataset/Permissions menu to only have one option - Dataset or File - if you only have the original ManageDatasetPermissions or the new ManageFilePermissions permission, respectively. (Existing roles which have both permissions after the PR still see a menu with the original two options.).
The new ManageFilePermissions permission shows up in the list for the admin and curator roles where that is shown (e.g. on the permission management dialogs).

**Comment by scolapasta** [comment]
Well, with the core code there would not actually be a UI change - only with a new role that only had one or the other permission.

I agree that I don't think we actually need docs (unless we do document already the permissions / roles that exist with that level of detail.

As far as release notes, do we normally create a note as part of the PR for a new feature like this if it doesn't actually require anything of the admin? Or does @djbrooke just add some detail for the new "feature" when he creates the notes?

**Comment by djbrooke** [comment]
@pdurbin @scolapasta I can add something to this branch, probably at the use case level. Not sure it raises to the level of a big chunk. :)

**Comment by qqmyers** [comment]
FWIW: I added a short release note - feel free to use/ignore the text there.

**Comment by kcondon** [comment]
Preliminary testing:
Basic smoketest works, admin and curator roles still can assign/revoke perms, ManageFilePermissions user can access ListRequests endpoint, non permitted user cannot.
Issues/questions:
1. Roles with only ManageFilePermissions or ManageDatasetPermissions cannot see edit->permissions in UI for dataset.
2. ListRequests endpoint throws log exception if don't provide api token, likely preexisting issue.
3. Was not able to get grantAccess or rejectAccess endpoints to work, likely syntax issue due to returning 404.

**Pull request merged**